Lets use packages

lshearer opened the pull request:
This PR brings in upstream changes that refactor the ESLint config package into a modular format and adapts the project to build an `eslint-config-hudl` [npm package](https://github.com/hudl/javascript/tree/lets-use-packages/packages/eslint-config-hudl).

The upstream merge also brought with it rule changes, which have been manually reverted in favor of a separate PR and discussion. Docs have been added that show the computed difference between the configs [before and after](https://github.com/hudl/javascript/blob/lets-use-packages/packages/eslint-config-hudl/docs/comparisons/before-packaging-refactor.md) this refactor PR. 

Additionally, a diff of [our styleguide and Airbnb's](https://github.com/hudl/javascript/blob/lets-use-packages/packages/eslint-config-hudl/docs/comparisons/with-airbnb-master.md) has been added to document variances between the two styleguides. By recording known, intentional differences between our styleguide and Airbnb's, we will be able to better manage the process of bringing in changes Airbnb has made to their styleguide.

Closes #14.

banhathome committed:
Rename jshintrc to .jshintrc

According to http://jshint.com/docs/, the "special file" to configure JSHint is named `.jshintrc`, not `jshintrc`.

amsardesai committed:
Added id-length documentation url

justjake committed:
Merge pull request #505 from amsardesai/add-id-length

Added id-length rule to eslintrc

justjake committed:
Merge pull request #491 from gilbox/react-terse-export-syntax

[react] inline export syntax

justjake committed:
Merge pull request #499 from honzi/honzi-patch-2

improve formatting of Chinese language links

sindrenm committed:
Add Hyper to the “in the wild” list in README

sindrenm committed:
Remove issues from how to add to “in the wild”

justjake committed:
Merge pull request #517 from hyperoslo/change/in-the-wild-instructions

Remove issues from how to add to “in the wild”

hshoff committed:
Merge pull request #516 from hyperoslo/add/hyperoslo-in-the-wild

Add Hyper to the “in the wild” list in README

Commit:
eslint-config-airbnb 0.0.9

Commit:
update readme for eslint-config-airbnb 0.0.9

Commit:
use eslint 1.3.0 (fixes #515)

DannyNemer committed:
Clarify proper type casting for Numbers

Clarifies when to use `Number` vs. `parseInt` for Number type casting.

Fixes: #523

ljharb committed:
Merge pull request #524 from DannyNemer/patch-1

Clarify proper type casting for Numbers

taion committed:
Use rules from eslint-config-defaults

taion committed:
Remove differences from eslint-config-defaults

taion committed:
Update for upstream dependencies

- react/jsx-quotes has been superseded by jsx-quotes
- eqeqeq now covers comparisons to null

justjake committed:
Merge pull request #526 from taion/eslint-config-defaults

Eslint config defaults

Commit:
create legacy.js to export es5 rules

Commit:
make exported rulesets extend relative to eslint-config-airbnb

Commit:
explain eslint-config-airbnb rule choices

justjake committed:
Merge pull request #527 from airbnb/jake/legacy-ruleset

Create eslint-config-airbnb/legacy to export ES5-only rules

Commit:
eslint-config-airbnb v0.1.0

justjake committed:
Merge pull request #528 from airbnb/jake/release-eslint-0.1.0

prepare eslint-config-airbnb v0.1.0

taion committed:
Clarify strict ESLint rule

ljharb committed:
Merge pull request #530 from taion/patch-1

Clarify strict ESLint rule

gmccue committed:
Add Ecosia to "In the Wild" section.

gmccue committed:
Fix link text.

hshoff committed:
Merge pull request #533 from ecosia/update-inthewild

Add Ecosia to "In the Wild" section.

cooperka committed:
Add reasoning for using double quotes in JSX

cooperka committed:
Add further reasoning for using double quotes in JSX

amsardesai committed:
Added version badge to eslint-config-airbnb

goatslacker committed:
Merge pull request #534 from cooperka/react-quote-reasoning

Add reasoning for using double quotes in JSX

ljharb committed:
Start to write a realistic "testing" section

Commit:
Add HubSpot to 'In The Wild' section

justjake committed:
Merge pull request #544 from HubSpot/in-the-wild

Add HubSpot to 'In The Wild' section

banhathome committed:
Add Section 18.8 - No padded blocks

This addresses #483

banhathome committed:
Update 17.2 to allow comments in the first line of a block

In #483, justjake said that we should update the style guide to allow
single line comments in the first line of a block.

banhathome committed:
Add examples with if statements to 18.8

majapw committed:
export default should not be inline with the class declaration

When there are static variables like `propTypes` and `defaultProps`, we shouldn't export them until we have actually set them onto the `Component`. This may change if the current ES7 proposal for static class properties gets accepts, and we update to support it.

adrienverge committed:
eslint: Add missing `space-before-keywords` rule

All examples in README.md seem to agree on avoiding missing whitespaces
before keywords such as:

    if (cond) {
    }else {       // no space before 'else'
    }

    try {
    }catch (e) {  // no space before 'catch'
    }

This patchs adds the `space-before-keywords` rule as an error (as is
`space-after-keywords` already).

justjake committed:
Merge pull request #554 from adrienverge/space-before-keywords

eslint: Add missing `space-before-keywords` rule

Commit:
Flag modification of variables that are declared using `const` keyword as an error.

ljharb committed:
Merge pull request #560 from rhbecker/update-eslint-config_no-const-assign

disallow reassignment of `const` variables

adrienverge committed:
eslint: Fix `no-mixed-spaces-and-tabs` turn off

Currently the way the rule is defined, mixing spaces and tabulations in
indentation is allowed. However, the comments and documents seems to say
the opposite.

This patchs turns the rule on so mixing spaces and tabs is NOT allowed.

Fixes: #539

ljharb committed:
Merge pull request #561 from adrienverge/issue-539

eslint: Enforce `no-mixed-spaces-and-tabs`

ReadmeCritic committed:
Update redirects

ljharb committed:
Merge pull request #562 from airbnb/ljharb/testing

Start to write a realistic "testing" section

hshoff committed:
Merge pull request #564 from frankenbot/master

Update redirects

hshoff committed:
[6.5] add backticks around stray eval. fixes #556

hshoff committed:
Merge pull request #566 from airbnb/fix-stray-eval

[6.5] add backticks around stray eval. fixes #556

hshoff committed:
Merge pull request #551 from majapw/patch-1

`export default` should not be inline with the class declaration

Commit:
[eslint] remove id length rule (too much nutrage)

justjake committed:
Merge pull request #569 from airbnb/jake/remove-id-length-rule

[eslint] remove id length rule (too much nutrage)

Commit:
eslint-config-airbnb 0.1.1

justjake committed:
Merge pull request #547 from chrisngobanh/padded-blocks

Prohibit Padded Blocks and Allow Comments in the First Line of a Block

ljharb committed:
[eslint] update `eslint`, `eslint-plugin-react`, `react`, `tape`.

ljharb committed:
[eslint] Remove `babel-eslint` dependency.

ljharb committed:
Add `eslint` as a peer dep, 1.0 or greater

ljharb committed:
Merge pull request #570 from airbnb/ljharb/update_eslint

[eslint config] update `eslint` to require 1.0 or greater

ljharb committed:
[eslint config] v1.0.0

joscha committed:
docs: babel-eslint not neede since 1.0.0 any more

ljharb committed:
Merge pull request #572 from joscha/patch-1

docs: babel-eslint not neede since 1.0.0 any more

Commit:
Add Evolution Gaming to 'In The Wild' section

Commit:
Organisation renamed

adrienverge committed:
[eslint] Enforce rule 18.5

Rule 18.5 [1] states that files shall end with a single newline
character. Until now, this was not checked because there was no such
option in eslint.

Since version 1.8.0, eslint provides the ability to do that. See pull
request on eslint [2] for details on implementation and usage. Let's use
it in airbnb eslint plugin now!

[1]: https://github.com/airbnb/javascript#18.5
[2]: https://github.com/eslint/eslint/pull/4266

justjake committed:
Merge pull request #578 from adrienverge/enforce-rule-18.5

[eslint] Enforce rule 18.5

Commit:
Bring back note about braces

Commit:
Use require.resolve to allow nested extend

Commit:
Avoid the word "mutate" while explaining `const`

goatslacker committed:
Merge pull request #584 from jonathanp/master

Avoid the word "mutate" while explaining `const`

ariporad committed:
Enable rest params!

ljharb committed:
Merge pull request #592 from ariporad/patch-1

[rules: es6] enable restParams

Commit:
[eslint config] v1.0.2

pigoz committed:
Add Section 18.9 - No spaces inside parentheses

See #593

pigoz committed:
Add Section 18.10 - No spaces inside brackets

See #593

pigoz committed:
Add Section 18.11 - Add spaces inside curly braces

Fixes #593

ljharb committed:
Merge pull request #594 from pigoz/parens-brackets-curlys

Make the spaces within parenthesis/brackes/curlys rules explicit

Commit:
Make 8.2 and 8.4 simpler and more explicit

Commit:
Clarify 8.4

Commit:
Fix typo

zwhitchcox committed:
Add facebook's react to "In the Wild"

I can't believe you guys didn't have that one! That's probably your best one.

Commit:
Adding jeopardybot to styleguide users.

ljharb committed:
Merge pull request #602 from kesne/jeopardy

Adding jeopardybot to styleguide users.

Commit:
Update README.md

Adding Bisk to the list of organizations

goatslacker committed:
Merge pull request #600 from zwhitchcox/patch-1

Add facebook's react to "In the Wild"

ljharb committed:
Merge pull request #604 from Bisk/master

Add Bisk to the list of organizations

ljharb committed:
Require a space between `function` and `(`, and `function` and the function's name, and disallow spaces between the function's name and `(`.

ljharb committed:
I abhor switch statements, but their indentation should be:

```js
switch (foo) {
  case 'bar':
    break;
}
```

not what the plugin currently requires:
```js
switch (foo) {
case: 'bar':
  break;
}
```

ljharb committed:
Merge pull request #606 from airbnb/ljharb/fix_switch_indent

[breaking] Fix switch statement indentation rule

ljharb committed:
Merge pull request #605 from airbnb/ljharb/fix_function_spacing

[breaking] Require function spacing: `function <optional name>(`

ljharb committed:
[eslint config] v2.0.0

 - separate changelog

JonAbrams committed:
Fix example for section 7.11

The missing newline caused the markdown parser to misinterpret the beginning and end of the code block.

goatslacker committed:
Merge pull request #607 from JonAbrams/patch-1

Fix example for section 7.11

goatslacker committed:
Documents corresponding eslint rules

goatslacker committed:
Merge pull request #608 from airbnb/document-eslint-rules

Documents corresponding eslint rules

winterbe committed:
Fix syntax highlighting in section 15

ljharb committed:
Merge pull request #615 from winterbe/fix15

Fix syntax highlighting in section 15

frenzzy committed:
Require space before/after arrow function's arrow (arrow-spacing)

Enable [arrow-spacing](http://eslint.org/docs/rules/arrow-spacing.html) rule, code with space before/after arrow function's arrow is easier to read.
```js
() => {};
(a) => {};
a => a;
() => {'\n'};
()=> {};     /*error Missing space before =>*/
() =>{};     /*error Missing space after =>*/
(a)=> {};    /*error Missing space before =>*/
(a) =>{};    /*error Missing space after =>*/
a =>a;       /*error Missing space after =>*/
a=> a;       /*error Missing space before =>*/
()=> {'\n'}; /*error Missing space before =>*/
() =>{'\n'}; /*error Missing space after =>*/
```

ljharb committed:
Merge pull request #617 from frenzzy/patch-1

[eslint config] [breaking] Require space before/after arrow function's arrow (`arrow-spacing`)

Commit:
[react] Including missing defaults to the react eslint

Commit:
Adding links to lint rules in react styleguide.

Commit:
Adding links to react rules for quick reference.

ljharb committed:
Merge pull request #618 from kesne/jgens/react-defaults

[eslint config] [react] [minor] Including missing defaults to the react eslint; enable react/jsx-quotes

Commit:
Cleaning up the react styleguide. Adding additional info in props.

Commit:
Removing file extensions. Fixing typos.

Commit:
Update README.md

Add OutBoxSoft to the list of organizations

ljharb committed:
Merge pull request #620 from OutBoxSoft/master

[Docs] Add OutBoxSoft to the list of organizations

ljharb committed:
[eslint config] enable `object-shorthand` rule.

Fixes #621.

justjake committed:
Merge pull request #582 from jkrems/jk-require-resolve

Use require.resolve to allow nested extend

ljharb committed:
[eslint config] [Dev Deps] update `eslint`

ljharb committed:
[eslint config] fix b1500397e1bba4120b10b850038bfff3177fda42

ljharb committed:
[eslint config] v2.1.0

billyjanitsch committed:
Remove deprecated react/jsx-quotes

ljharb committed:
Merge pull request #622 from billyjanitsch/jsx-quotes

[eslint config] [fix] Remove deprecated react/jsx-quotes

ljharb committed:
[eslint config] add rule link and defaults to `jsx-quotes` rule

ljharb committed:
v2.1.1

trshafer committed:
update best-practices config to prevent parameter object manipulation
added good/bad examples of parameter mutation to the readme

trshafer committed:
update examples for style
add link to no-param-reassign rule documentation

trshafer committed:
added newlines after each function

ljharb committed:
Merge pull request #627 from amplii/no-param-reassign-with-props

[eslint config] [breaking] update best-practices config to prevent parameter object manipulation

ljharb committed:
[eslint config] [Dev Deps] update `eslint-plugin-react`

ljharb committed:
[eslint config] [minor] enable react/prefer-es6-class rule

ljharb committed:
[eslint config] [breaking] enable `quote-props` rule.

lencioni committed:
[eslint config] [minor] Sort react/prefer-es6-class alphabetically

All of the other rules in this file are in alphabetical order, but this
one was added by c98990c0f out of order at the end. Keeping these in
alphabetical order will help developers find the rules that they are
looking for.

lencioni committed:
[eslint-config] [minor] Enable react/no-is-mounted rule

isMounted is an anti-pattern [0], is not available when using ES6
classes, and is on its way to being officially deprecated.
eslint-plugin-react recently added the react/no-is-mounted rule in
3.12.0 that prevents its use.

[0]: https://facebook.github.io/react/blog/2015/12/16/ismounted-antipattern.html

Finishes #633

ljharb committed:
Merge pull request #634 from lencioni/sort-prefer-es6-class

[eslint config] Sort react/prefer-es6-class alphabetically

ljharb committed:
Merge pull request #635 from lencioni/no-is-mounted

[eslint-config] [breaking] Enable react/no-is-mounted rule

lencioni committed:
[readme] Clean up isMounted section in react readme

When I added this section in f2dc5043a I didn't do a very good job of
following the style used in other parts of this repo. This commit
remedies that.

lencioni committed:
[readme] Add backticks around code items in react readme

I noticed a number of places in this document where code was being
referenced but it was not marked as such. Adding backticks will instruct
the markdown parser to format these bits as code, which should improve
the readability of this document.

lencioni committed:
[readme] Add missing heading to react TOC

I noticed that this heading wasn't listed in the table of contents.
Adding it will help people see at-a-glance what they might find in this
document.

lencioni committed:
[readme] Use consistent whitespace in react readme

I noticed that some newlines and indentation were a little inconsistent
in this file, so I decided to smooth things out a bit. This should help
people who decide to modify this document in the future. When
determining which is the "right" way to format these things, I decided
to use the style used by base readme as a guide.

evenfrost committed:
Permit strict mode

Permit strict mode for legacy configuration as it is not intended to be used with Babel.

ljharb committed:
Merge pull request #636 from lencioni/react-readme-cleanup

[Docs] React readme cleanup

ljharb committed:
Merge pull request #576 from evolution-gaming/in-the-wild

[in the wild] Add Evolution Gaming

Commit:
Removing name prefix rules.

ljharb committed:
Merge pull request #536 from amsardesai/add-version-badge

[eslint config] [Docs] Add version badge

ljharb committed:
Merge pull request #421 from chrisngobanh/jshintrc-fix

Rename jshintrc to .jshintrc

ljharb committed:
Merge pull request #619 from kesne/jgens/react-cleanup

[eslint config] [breaking] [react] Minor cleanup for the React styleguide

banhathome committed:
Add section 18.12 on line length

banhathome committed:
Update .jshintrc with new line length

banhathome committed:
Enable ESLint rule on limiting maxiumum length of line (max-len)

I moved the rule from legacy.js to style.js and enabled it

banhathome committed:
Reword section 6.2 to recognize the line length

tikotzky committed:
add missing rules from eslint-plugin-react

tikotzky committed:
add default config to all rules which take a config

ljharb committed:
Merge pull request #581 from tikotzky/update-eslint-react

[eslint config] [react] add missing rules from eslint-plugin-react (enforcing where necessary)

evenfrost committed:
Preserve strict rule for default export

banhathome committed:
Address comments in #639

[eslint] Add comment above `max-len` rule with link to its docs
[eslint] Change tab width for `max-len` rule from 4 to 2
[eslint] Replace double quotes around `max-len` with single quotes
[eslint] Use object form of `max-len` and include all of the options

ljharb committed:
Merge pull request #639 from chrisngobanh/max-len

[eslint config] [breaking] Define a max line length of 100 characters

ljharb committed:
Merge pull request #637 from evenfrost/master

[eslint config] Permit strict mode in "legacy" config

SimenB committed:
Allow multiple stateless components in a single file

AlicanC committed:
Make reassigning a separate rule.

ljharb committed:
Merge pull request #643 from AlicanC/split-never-mutate-parameters

[Docs] Make no-param-reassign a separate section.

AlicanC committed:
Fix indentation and spacing.

yuyokk committed:
Add missing back tick

ljharb committed:
Merge pull request #648 from yuyokk/patch-1

[Docs] Add missing back tick

ljharb committed:
Merge pull request #647 from AlicanC/cleanup

[Docs] Cleanup markdown indentation

AlicanC committed:
Add ESLint link for "Never mutate parameters." rule.

ljharb committed:
Merge pull request #649 from AlicanC/never-mutate-parameters-link

[Docs] Add eslint link for "Never mutate parameters." rule

Commit:
Fix max-len rule definition

The current definition is valid only in eslint@2.x.x, but this package has a dependency on eslint@1.10.3 which doesn't have all those options yet.

ljharb committed:
Merge pull request #651 from gergoerdosi/max-len

[Fix] fix max-len rule definition for eslint 1.x

strburst committed:
Link newer ESLint rule (namely comma-dangle)

The "no-comma-dangle" rule was deprecated in favor or "comma-dangle" in
ESLint v1.0.

See http://eslint.org/docs/rules/no-comma-dangle.html,
http://eslint.org/docs/rules/comma-dangle, and line 4 of
packages/eslint-config-airbnb/rules/errors.js.

ljharb committed:
Merge pull request #652 from strburst/link-comma-dangle

[docs] Link newer ESLint rule (no-comma-dangle → comma-dangle)

Commit:
missing 's', create(s) under Method

ljharb committed:
Merge pull request #655 from awitherow/master

[guide] missing 's', create(s) under Method

Commit:
Require parentheses

As suggested by @ljharb in https://github.com/airbnb/javascript/pull/579#issuecomment-166981740

ljharb committed:
Merge pull request #579 from tomekwi/fix-braces-text

[guide] Bring back note about braces

ljharb committed:
Merge pull request #632 from airbnb/ljharb/quote_props

[eslint config] [breaking] enable `quote-props` rule.

eddiemonge committed:
assign variables: give a reason for the assignment

The variable assignment wasn't necessary. This gives a reason to have it there

ljharb committed:
v3.0.0

ljharb committed:
Fix a bug introduced in f2afce75e949eacd1ba64e89b863105828976f20 / #581

mccambridge committed:
clarity in destructuring code example

readme

ljharb committed:
Merge pull request #657 from mccambridge/master

[guide] clarity in destructuring code example

jonny-no1 committed:
Fixed typo in URL

ljharb committed:
Merge pull request #658 from foaly-nr1/patch-1

[guide] Fixed typo in URL

ljharb committed:
[eslint config] [fix] because we use babel, keywords should not be quoted.

ljharb committed:
v3.0.1

silvenon committed:
Link to the eslint rule for radix in parseInt

ljharb committed:
Merge pull request #663 from silvenon/patch-1

[guide] Link to the eslint rule for radix in parseInt

silvenon committed:
Ignore URLs in max-len

goatslacker committed:
Merge pull request #664 from silvenon/patch-1

Ignore URLs in max-len

ljharb committed:
v3.0.2

ljharb committed:
Merge pull request #575 from SimenB/patch-1

[eslint config] [minor] Allow multiple stateless components in a single file

ljharb committed:
[guide] clarify function spacing in 18.3

Per #668.

ljharb committed:
v3.1.0

ljharb committed:
[eslint config] [breaking] require outer IIFE wrapping; flesh out guide section.

There was lots of discussion [here](https://github.com/airbnb/javascript/issues/21#issuecomment-10203921), but now that we have both a modern build system and an eslint rule requiring terminating semicolons, the concerns with the “crockford” style no longer apply.

simono committed:
7.3: Add Link to eslint rule.

simono committed:
22.3: Add Link to eslint rule.

ljharb committed:
Merge pull request #675 from simono/patch-1

[guide] 7.3: Add Link to eslint rule.

simono committed:
Add `prefer-arrow-callback` to ES6 Rules.

This prefers arrow functions as described in https://github.com/airbnb/javascript#8.1

ljharb committed:
Merge pull request #677 from simono/patch-3

[eslint config] [minor] Add `prefer-arrow-callback` to ES6 rules (to match the guide)

pgeiss committed:
Clearly show that [] is truthy

simono committed:
Add missing ES6 rules to ESLint config.

`arrow-body-style` (8.2) and `prefer-template` (6.4) are described in
the Style Guide.

jshen212 committed:
Fix case type on line 3 comment

Update case type to resemble other comments

ljharb committed:
[Tests] run `npm run lint` as part of tests; fix errors.

hat tip to https://github.com/airbnb/javascript/pull/678#issuecomment-172017623

prayagverma committed:
Update license year range to 2016

ljharb committed:
[Tests] use `parallelshell` to parallelize npm run-scripts

hzoo committed:
Document corresponding jscs rules

hzoo committed:
remove specific rule values in text

Commit:
fix typo

ljharb committed:
[docs] remove trailing whitespace

sveinhal committed:
arrow-parens

ljharb committed:
[guide] fix a few IIFE examples

hshoff committed:
Merge pull request #683 from kvzivn/patch-1

fix typo

Commit:
Favour stateless functions over classes when there's no state

Commit:
Remove space from jsx literal and add space to function definition

According to PR comments.

Commit:
Fix PR comments

- Space between function name and args;
- use `render(){}` syntax in object literal;
- Fix link to section;

brettjurgens committed:
add avant to users

ljharb committed:
Merge pull request #688 from joaomilho/patch-2

[guide] Favor stateless functions over classes when there's no state

rickyeh committed:
Fixed minor capitalization change in Hoisting

Not sure if it's an actual fix, but just noticed this was the only comment that had capitalization that wasn't a proper noun

Commit:
add WeBox Studio to the organizations list

hshoff committed:
Merge pull request #679 from jshen212/patch-1

Fix case type on line 3 comment

ljharb committed:
v4.0.0

danielrussellLA committed:
Fixed capitalization

Bitshift changed to bitshift

hshoff committed:
[rule links] reduce visual clutter

hshoff committed:
[react][rule links] condense lines

hshoff committed:
[react][rule links] add eslint labels back

hshoff committed:
[rule links] add labels back

hshoff committed:
Merge pull request #695 from airbnb/move-rules

[rule links] reduce visual clutter

apaleslimghost committed:
Remove language interpretable as an ableist slur

ljharb committed:
[guide] [react] add a note preferring normal functions for functional stateless components.

ethanrubio committed:
Fixed capitalization

Fixed capitalization of ES6 to match the rest of the README.

LarryBattle committed:
Added names to invalid function declarations

fredrikolovsson committed:
Edit sample code in 24.1 to follow rule 8.1

Replaced function expression with arrow function notation.

marksanghoonkim committed:
Fixed capitalization and comment spacing

Fixed line 1219 to have a lowercase 'a' to be consistent with the style guide.
Fixed line 1831 to only have one space between the // and the =>

prayagverma committed:
Minor typo

Remove extra dot

lshearer committed:
Merge remote-tracking branch 'airbnb/master' into lets-use-packages

# Conflicts:
#	README.md
#	linters/.eslintrc
#	linters/.jshintrc
#	react/README.md

lshearer committed:
Adjust updates and rename airbnb -> hudl

- Incorporate updates and adjust for publishing as a hudl package
- Adding in some gulp task stubs to expand on

lshearer committed:
use simple tests for now

lshearer committed:
Add eslint-teamcity reporter

lshearer committed:
Use more appropriate settings for node code

lshearer committed:
Add some comparison generation

lshearer committed:
Use ins/del instead of span

lshearer committed:
Allow defining known differences

lshearer committed:
Regenerate diff

lshearer committed:
Flesh out comparison docs and testing around them

lshearer committed:
Fix relative paths

lshearer committed:
Update Readme and point to comparison docs

lshearer committed:
Fix rule grouping

lshearer committed:
Fix linting errors

lshearer committed:
Add mocha-teamcity-reporter, fix docs

lshearer committed:
Filter out removed disabled rules

lshearer committed:
Change new rules back to current standard

- We will be able to revisit this later by comparing by Airbnb’s
current version and then recording known differences.

lshearer committed:
Normalize path slashes for windows

lshearer committed:
Add slash package

lshearer committed:
Fix original tests

codeimpossible commented:
This is epic.

blainekasten commented on README.md:
````diff
@@ -402,7 +424,7 @@ Forked from [AirBNB's style guide](https://github.com/airbnb/javascript)
     ```
 
-  - [6.2](#6.2) <a name='6.2'></a> Strings longer than 80 characters should be written across multiple lines using string concatenation.
+  - [6.2](#6.2) <a name='6.2'></a> Strings that cause the line to go over 100 characters should be written across multiple lines using string concatenation.
````
do we want to encourage string templates instead of concatenation? also isn't our linter set up for 160 chars?

blainekasten commented on README.md:
````diff
@@ -566,7 +592,21 @@ Forked from [AirBNB's style guide](https://github.com/airbnb/javascript)
+    function handleThings(name, opts = {}) {
+      // ...
+    }
+    ```
````
I'm not so sure about this one. Sometimes consistency can make more sense. For example, say I have an external API that just calls an internal one, and the internal one defaults things. 

ie:

``` js
@public
function foo(bar, baz) {
  internalFoo(bar, baz);
}

@private
function internalFoo(bar = 1, baz) {
  // things
}
```

In situations like these, I don't want to flip the signatures, because then I would just get confused about which one goes where all the time. We have implementations like this in Video Interface. So I don't think it makes much sense, nor is really necassary.

colevoss replied:
Having default params last is really smart. Otherwise they are kinda pointless in certain situations. You might want to pass only of of the arguments to the function so you can use the default of the other argument. 

In the example above, having the first argument defaulted and wanting to take advantage of it would look like this

``` javascript
function handleThings(opts = {}, name) {
  // ...
}

// use it
handleThings({}, 'name string');
```

That kind of defeats the purpose of the default param.

Instead, you would just have the following with defaulted params as last.

``` javascript
function handleThings(name, opts = {}) {
  // ...
}

// use it
handleThings(name)
```

blainekasten commented on README.md:
```diff
@@ -578,11 +618,65 @@ Forked from [AirBNB's style guide](https://github.com/airbnb/javascript)
+
+    // good
+    function f2(obj) {
+      const key = Object.prototype.hasOwnProperty.call(obj, 'key') ? obj.key : 1;
```
these two examples are different. One is a setter, the other is a reader. To mirror, you would want `f2` to be something like

``` js
function f2(obj) {
  return merge(obj, { key: 1 })
}
```

colevoss replied:
I agree that these two functions don't really reflect the same functionality but I don't think @blainekasten's example is what you would want in `f2` either. Something like

``` javascript
function f2(obj) {
  return Object.assign({}, obj, {key: 1});
}
```

would better reflect not manipulating the passed in parameter.

blainekasten commented on README.md:
```diff
@@ -591,16 +685,18 @@ Forked from [AirBNB's style guide](https://github.com/airbnb/javascript)
     [1, 2, 3].map((x) => {
-      return x * x;
+      const y = x + 1;
+      return x * y;
```
fwiw: The previous example was fine as it wasn't actually changing the `x` variable.

blainekasten commented on README.md:
```diff
@@ -608,12 +704,68 @@ Forked from [AirBNB's style guide](https://github.com/airbnb/javascript)
+    [1, 2, 3].map(number => {
+      const nextNumber = number + 1;
+      `A string containing the ${nextNumber}.`;
+    });
```
rather than `// bad` this is more like `// wont work`

blainekasten commented on README.md:
```diff
@@ -608,12 +704,68 @@ Forked from [AirBNB's style guide](https://github.com/airbnb/javascript)
+    [1, 2, 3].map(number => (
+      `As time went by, the string containing the ${number} became much ` +
+      'longer. So we needed to break it over multiple lines.'
+    ));
```
do we want to just use template strings for multiple lines?

lshearer committed:
Add babel-lint parser for react config

- Add babel-lint parser for only react (default) config
- Add some cross-links in docs

lshearer committed:
Re-enable all tests

blainekasten commented on README.md:
```diff
@@ -608,12 +704,68 @@ Forked from [AirBNB's style guide](https://github.com/airbnb/javascript)
-    }, 0);
+    [1, 2, 3].map(number => (
+      `A long string with the ${number}. It’s so long that we’ve broken it ` +
+      'over multiple lines!'
```
this wouldn't work as it doesn't actually return anything.

blainekasten replied:
I'm not sure I'm a fan of this. It's deceiving since `(` look similar to `{`. 
I just assumed this would't work because I didn't think it was implicitly returning..

lshearer commented:
The focus of this PR is really the packaging of the current ESLint config, so I'll revert the markdown changes so they can be discussed (in addition to the reverted Airbnb ESLint config changes) separately.

lshearer committed:
Revert style guide changes and fix linters/README

lshearer committed:
Add node config example

Adding an example of a node configuration to the README.

Closes #17.

lshearer commented:
The only outstanding comments are related to the now-reverted changes to the markdown style guide pages, so I'm considering this good-to-go.
